lib/unstrctrd.ml

```diff
@@ -82,38 +82,54 @@ let to_utf_8_string ?(rep= Uutf.u_rep) lst =
     | [] -> if stack = 0 then Ok (List.rev acc) else error_msgf "Non-terminating comment"
     | `Uchar uchar as value :: r ->
       ( match Uchar.to_int uchar with
+        | 0x22 (* '"' *) ->
+          ( match escaped, quoted_string, stack with
+          | true,  true,  0 -> go stack ~escaped:false ~quoted_string (value :: acc) r
+          | true,  false, 0 -> go stack ~escaped:false ~quoted_string (value :: acc) r
+          | false, true,  0 -> go stack ~escaped ~quoted_string:false (value :: acc) r
+          | false, false, 0 -> go stack ~escaped:false ~quoted_string:true (value :: acc) r
+          | true,  true,  _ -> go stack ~escaped:false ~quoted_string acc r
+          | true,  false, _ -> go stack ~escaped:false ~quoted_string acc r
+          | false, true,  _ -> go stack ~escaped ~quoted_string:false acc r
+          | false, false, _ -> go stack ~escaped ~quoted_string:true  acc r )
         | 0x28 (* '(' *) ->
-          if not escaped
-          then go (succ stack) false acc r
-          else
-            ( if stack > 0
-              then go stack false acc r
-              else go stack false (value :: acc) r )
+          ( match escaped, quoted_string, stack with
+          | true,  true,  0 -> go stack ~escaped:false ~quoted_string (value :: acc) r
+          | true,  false, 0 -> go stack ~escaped:false ~quoted_string (value :: acc) r
+          | true,  true,  _ -> go stack ~escaped:false ~quoted_string acc r
+          | true,  false, _ -> go stack ~escaped:false ~quoted_string acc r
+          | false, true , 0 -> go 0 ~escaped ~quoted_string (value :: acc) r
+          | false, true,  _ -> go stack ~escaped ~quoted_string acc r
+          | false, false, n -> go (succ n) ~escaped ~quoted_string acc r )
         | 0x29 (* ')' *) ->
-          if not escaped
-          then go (pred stack) false acc r
-          else
-            ( if stack > 0
-              then go stack false acc r
-              else go stack false (value :: acc) r )
+          ( match escaped, quoted_string, stack with
+          | true,  true,  0 -> go stack ~escaped:false ~quoted_string (value :: acc) r
+          | true,  false, 0 -> go stack ~escaped:false ~quoted_string (value :: acc) r
+          | true,  true,  _ -> go stack ~escaped:false ~quoted_string acc r
+          | true,  false, _ -> go stack ~escaped:false ~quoted_string acc r
+          | false, true,  0 -> go 0 ~escaped ~quoted_string (value :: acc) r
+          | false, true,  _ -> go stack ~escaped ~quoted_string acc r
+          | false, false, n -> go (pred n) ~escaped ~quoted_string acc r )
         | 0x5c (* '\' *) ->
-          if not escaped
-          then go stack true acc r
-          else
-            ( if stack > 0
-              then go stack false acc r
-              else go stack false (value :: acc) r )
+          ( match escaped, quoted_string, stack with
+          | true,  true,  _ -> go stack ~escaped:false ~quoted_string (value :: acc) r
```

Comment:
No reason here to not consider the case you are inside a comment. The string ```"(Hi! \\\\)Hello\r\n"``` without comment should be ```"Hello"```.